File: coding_skill_cpp/04_callback_pattern.cpp
/* C++ Callback - std::function과 람다 */
#include <iostream>
#include <functional>
#include <vector>
using namespace std;

// 콜백 타입 정의
using Callback = function<void(int)>;
using DataCallback = function<void(const string&)>;

class Button {
    Callback onClick;
public:
    void setOnClick(Callback cb) { onClick = cb; }
    void click() {
        cout << "[Button] 클릭!" << endl;
        if (onClick) onClick(1);
    }
};

class Observable {
    vector<Callback> observers;
public:
    void attach(Callback cb) { observers.push_back(cb); }
    void notify(int data) {
        cout << "[Observable] 통지: " << observers.size() << "명" << endl;
        for (auto& obs : observers) obs(data);
    }
};

int main() {
    cout << "\n=== C++ Callback Pattern ===" << endl;
    
    // 람다 콜백
    Button btn;
    btn.setOnClick([](int x) { cout << "  → 콜백 호출: " << x << endl; });
    btn.click();
    
    // 다중 콜백
    Observable obs;
    obs.attach([](int x) { cout << "  → Observer 1: " << x << endl; });
    obs.attach([](int x) { cout << "  → Observer 2: " << x * 2 << endl; });
    obs.notify(10);
    
    return 0;
}

File: coding_skill_py/04_callback_pattern.py
"""Python Callback - 일급 함수와 람다"""
from typing import Callable, List

class Button:
    def __init__(self):
        self.on_click = None
    
    def set_on_click(self, callback: Callable):
        self.on_click = callback
    
    def click(self):
        print("[Button] 클릭!")
        if self.on_click:
            self.on_click(1)

class Observable:
    def __init__(self):
        self.observers: List[Callable] = []
    
    def attach(self, callback: Callable):
        self.observers.append(callback)
    
    def notify(self, data):
        print(f"[Observable] 통지: {len(self.observers)}명")
        for obs in self.observers:
            obs(data)

def main():
    print("\n=== Python Callback Pattern ===")
    
    # 람다 콜백
    btn = Button()
    btn.set_on_click(lambda x: print(f"  → 콜백 호출: {x}"))
    btn.click()
    
    # 다중 콜백
    obs = Observable()
    obs.attach(lambda x: print(f"  → Observer 1: {x}"))
    obs.attach(lambda x: print(f"  → Observer 2: {x * 2}"))
    obs.notify(10)

if __name__ == "__main__":
    main()
